ARCHITECTURE.md

```diff
@@ -0,0 +1,231 @@
+# Istio operator code overview
+
+## Introduction
+
+This document covers primarily the code, with some background on how the design maps to it.
+See the 
+[design doc](https://docs.google.com/document/d/11j9ZtYWNWnxQYnZy8ayZav1FMwTH6F6z6fkDYZ7V298/edit#heading=h.qex63c29z2to)
+for a more complete design description. The operator code is divided roughly into five areas:
+
+1. [New API](The_new_API) (`IstioControlPlaneSpec`) and related infrastructure, which is expressed as a
+[proto](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto) and
+compiled to [Go
+structs](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.pb.go).
+`IstioControlPlaneSpec` has pass-through fields to the Helm values.yaml API, but these are additionally validated through
+a [schema](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/values_types.go). 
+1. [Controller](K8s_controller) code. The code comprises the K8s listener, webhook and logic for reconciling the cluster
+to an `IstioControlPlaneSpec` CR. 
+1. [Manifest creation](Manifest_creation) code. User settings are overlaid on top of the
+selected profile values and passed to a renderer in the Helm library to create manifests. Further customization on the
+created manifests can be done through overlays. 
+1. [CLI](CLI) code. CLI code shares the `IstioControlPlaneSpec` API with
+the controller, but allows manifests to be generated and optionally applied from the command line without the need to
+run a privileged controller in the cluster. 
+1. [Migration tools](Migration_tools). The migration tools are intended to
+automate configuration migration from Helm to the operator.
+
+The operator code uses the new Helm charts in the [istio/installer](https://github.com/istio/installer) repo. It is not
+compatible with the older charts in [istio/istio](https://github.com/istio/istio/tree/master/install/kubernetes/helm).
+See the istio/installer repo for details about the new charts and why they were created. Briefly, the new charts
+are intended to support production ready deployments of Istio that follow best practises like canarying for upgrade.
+
+## Terminology
+
+Throughout the document, the following terms are used:
+
+- `IstioControlPlaneSpec`: The API directly defined in the
+[IstioControlPlaneSpec proto](https://github.com/istio/operator/blob/master/pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto),
+including feature and component groupings, namespaces and enablement, and per-component K8s settings. The term new API
```

Comment:
new API -> "new API" or _new API_

Reply:
I've removed all references to the old and new API.